src/replication.c

```diff
@@ -500,6 +499,54 @@ trigger_rebuild(spec_t *spec)
 	}
 }
 
+static bool
+is_replica_new_or_familar(spec_t *spec, replica_t *new_replica)
+{
+	known_replica_t *kr = NULL;
+	int familiar_replicas = 0;
+	bool newly_connected = false, found = false;
+	replica_t *old_replica = NULL;
+
+	ASSERT(MTX_LOCKED(&spec->rq_mtx));
+
+	TAILQ_FOREACH(kr, &spec->identified_replica, next) {
+		familiar_replicas++;
+		if (kr->zvol_guid == new_replica->zvol_guid) {
+			found = true;
+			if (!kr->is_connected) {
+				newly_connected = true;
+				kr->is_connected = true;
+			}
+			break;
+		}
+	}
+
+	if (!found && familiar_replicas < spec->replication_factor) {
+		kr = malloc(sizeof (known_replica_t));
+		kr->zvol_guid = new_replica->zvol_guid;
+		kr->is_connected = true;
+		newly_connected = true;
+		TAILQ_INSERT_TAIL(&spec->identified_replica, kr, next);
+	}
+
+	/*
+	 * If the target has already an active session with this replica
+	 * then the target will not allow new session
+	 */
+	if (found && !newly_connected) {
+		REPLICA_ERRLOG("replica(%lu) has already an active session\n",
+		    kr->zvol_guid);
+		/*
+		 * handle_mgmt_conn_error will update identified_replica status
```

Comment:
thats a very good explanation @mynktl ..
I'm thinking whether same can apply to other error scenarios in update_replica_entry after setting zvol_guid.
We need to move this check for is_replica_new_or_familiar right after setting zvol_guid in update_replica_entry.
Sorry to ask one more change: can you change is_replica_new_or_familiar to is_replica_newly_connected?